### DISCUSSION TITLE
validate config values of a module in setup

### DISCUSSION THREAD

**Discussion by jonasem** [nuxt/nuxt#26858] [Questions] [answered]
Hello, I have some runtime config values that should be validated before the app starts.

The problem is that the runtimeConfig for the module is not loaded during setup of the module itself. 

My impression from reading the docs is that a good practice is to augment default runtime options with env using `defu`, but that implies that `nuxt.options.runtimeConfig[configKey]` actually is loaded, which right now, seems to be empty.

Just tried the code below on a clean new nuxt app using the latest nuxt version 3.11.2.

Please advice, **how do I load the runtimeConfig during module setup?**

```typescript
import { defineNuxtModule } from "@nuxt/kit";

export default defineNuxtModule<{
  bar: string;
  runtimeConfig: { foo: string };
}>({
  meta: {
    name: "some-name",
    version: "1.0.0",
    configKey: "moduleName",
  },
  // Default configuration options of the Nuxt module
  defaults: {
    bar: "default bar",

    runtimeConfig: {
      foo: "default foo",
    },
  },

  setup(options, nuxt) {
    const config = nuxt.options.runtimeConfig;
    const env = Object.entries(process.env).filter(([k]) => k.startsWith("NUXT"));

    console.log(config, env);
    // logs { public: {}, app: { baseURL: '/', buildAssetsDir: '/_nuxt/', cdnURL: '' } } [ [ 'NUXT_MODULE_NAME_FOO', 'FOOENV' ] ]

    // @ts-expect-error
    if (config.moduleName?.foo === "foo default") {
      throw new Error("Invalid config, override with env");
    }
  },
});

```

**Comment by dsvgl** [comment]
It's a bit confusing that you named one of the module options `runtimeConfig`. Not sure what's the problem is here and what do you expect as outcome. Could you maybe share a sample/configured nuxt.config.ts that uses this module and describe happy and bad paths?

**Comment by jonasem** [comment]
Thanks for your reponse! I will try to clarify :)

The runtime config will be populated with values from the environment. And I'm following the pattern from the module author guide here https://nuxt.com/docs/guide/going-further/modules#exposing-options-to-runtime 

The expected outcome is that `nuxt.options.runtimeConfig.moduleName.foo` has the value set in the environment variable named `NUXT_MODULE_NAME_FOO`. When using a component later in the runtime, the value is indeed there, but not during setup.

So, for me, the example from the module author guide seems misleading, or there is a bug. It makes no sense to send an empty object into `defu`, or, nuxt has failed to populate the runtime options into the runtime config before running `setup`. I haven't found a spec saying that it should, but the example implies that it should.

I hope this helps and clears up my question!

**Reply by dsvgl** [reply, marked as answer]
Ok. I got it working: https://stackblitz.com/edit/github-pgdfdr?file=src%2Fmodule.ts
run these commands:

```
npm install
npm run dev:prepare
npm run dev
```

I think the problem is that the specially named env vars mentioned [here](https://nuxt.com/docs/guide/going-further/runtime-config#environment-variables) don't seem to work in the build step.

So, if having this in nuxt.config:

```js
runtimeConfig: {
  moduleNameFoo: '',
},
```

Then staring your app via `NUXT_MODULE_NAME_FOO=something npm run dev`. Logging out `nuxt.options.runtimeConfig.moduleNameFoo` inside the module one can see that it's just that (same) empty string.

Doing this will make the var present for the module:

```js
runtimeConfig: {
  moduleNameFoo: process.env.NUXT_MODULE_NAME_FOO || '',
},
```

This behavior makes sense. The automagic of reading env vars and adjust runtimeConfig does not happen in the build hook. All build related things needs to work with "raw" process.env.xxx. The nuxt.config itself also this "limitation". One cannot access things inside the thing that sets it up :)

Hope this helps.

BTW1: I set `NUXT_MODULE_NAME_FOO=...` in front of the dev script of the main package.json.
BTW2: I used the Nuxt 3 module starter from here https://nuxt.new/

**Reply by jonasem** [reply]
Thank you so much for your detailed reply! I now fully understand, and I see that for a module to validate it's own runtime-config, it cant be done during setup unless looking directly at `process.env`. Not ideal as that will break if the module name changes, but it works.
